include/clap/ext/draft/location.h

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include "../../color.h"
+#include "../../plugin.h"
+#include "../../string-sizes.h"
+
+// This extension allows a host to tell the plugin more about its position
+// within a project or session.
+
+static CLAP_CONSTEXPR const char CLAP_EXT_LOCATION[] = "clap.location/1";
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+enum {
+   // Represents a document/project/session.
+   CLAP_PLUGIN_LOCATION_PROJECT = 1,
+
+   // Represents a group of tracks.
+   // It can contain both track groups, tracks and devices (post processing).
+   // The first device within a track group has the index of
+   // the last track or track group within this group + 1.
+   CLAP_PLUGIN_LOCATION_TRACK_GROUP = 2,
+
+   // Represents a single tracks. It contains devices (serial).
+   CLAP_PLUGIN_LOCATION_TRACK = 3,
+
+   // Represents a single device.
+   // It contains other nested device chains.
+   CLAP_PLUGIN_LOCATION_DEVICE = 4,
+
+   // Represents a nested device chain (serial).
+   // Its parent must be a device.
+   // It contains other devices.
+   CLAP_PLUGIN_LOCATION_NESTED_DEVICE_CHAIN = 5,
+};
+
+typedef struct clap_plugin_location_element {
+   // Kind of the element, must be one of the CLAP_PLUGIN_LOCATION_* values.
+   uint32_t kind;
+
+   // Index within the parent element.
+   // Set to 0 if irrelevant.
+   uint32_t index;
```

Comment:
So `index` does *not* refer to the array index of this element's parent element within the array passed to `set_location`?

I could be completely wrong, but it seems like a simpler way of flattening the tree would be to specify that the order of elements in `set_location`'s `path` array is the same order that elements of would be visited in a depth-first traversal of the tree. Then each `index` would be the array index of the direct parent. And since the root element does not have a parent, its `index` could be the index of the plugin device itself.